README
NAME
    Catalyst::TraitFor::Controller::DBIC::DoesPaging - Helps you paginate,
    search, sort, and more easily using DBIx::Class

VERSION
    version 1.001002

SYNOPSIS
     package MyApp::Controller::Foo;
     use Moose;
     BEGIN { extends 'Catalyst::Controller' }
     with 'Catalyst::TraitFor::Controller::DBIC::DoesPaging';

     sub people {
        my ($self, $c) = @_;
        my $people = $self->page_and_sort(
           $self->search( $self->model('DB::People') )
        );
        # ...
     }

DESCRIPTION
    This module helps you to map various DBIx::Class features to CGI
    parameters. For the most part that means it will help you search, sort,
    and paginate with a minimum of effort and thought.

METHODS
    All methods take the context and a ResultSet as their arguments. All
    methods return a ResultSet.

  page_and_sort
     my $result = $self->page_and_sort($c, $c->model('DB::Foo'));

    This is a helper method that will first "sort" your data and then
    "paginate" it.

  paginate
     my $result = $self->paginate($c, $c->model('DB::Foo'));

    Paginates the passed in resultset based on the following CGI parameters:

     start - first row to display
     limit - amount of rows per page

  search
     my $searched_rs = $self->search($c, $c->model('DB::Foo'));

    If the $resultset has a "controller_search" method it will call that
    method on the passed in resultset with all of the CGI parameters. I like
    to have this method look something like the following:

     # Base search dispatcher, defined in MyApp::Schema::ResultSet
     sub _build_search {
        my $self           = shift;
        my $dispatch_table = shift;
        my $q              = shift;

        my %search = ();
        my %meta   = ();

        foreach ( keys %{$q} ) {
           if ( my $fn = $dispatch_table->{$_} and $q->{$_} ) {
              my ( $tmp_search, $tmp_meta ) = $fn->( $q->{$_} );
              %search = ( %search, %{$tmp_search||{}} );
              %meta   = ( %meta,   %{$tmp_meta||{}} );
           }
        }

        return $self->search(\%search, \%meta);
     }

     # search method in specific resultset
     sub controller_search {
        my $self   = shift;
        my $params = shift;
        return $self->_build_search({
           status => sub {
              return { 'repair_order_status' => shift }, {};
           },
           part_id => sub {
              return {
                 'lineitems.part_id' => { -like => q{%}.shift( @_ ).q{%} }
              }, { join => 'lineitems' };
           },
        },$params);
     }

    If the "controller_search" method does not exist, this method will call
    "simple_search" instead.

  sort
     my $result = $self->sort($c, $c->model('DB::Foo'));

    Exactly the same as search, except calls "controller_sort" or
    "simple_sort". Here is how I use it:

     # Base sort dispatcher, defined in MyApp::Schema::ResultSet
     sub _build_sort {
        my $self = shift;
        my $dispatch_table = shift;
        my $default = shift;
        my $q = shift;

        my %search = ();
        my %meta   = ();

        my $direction = $q->{dir};
        my $sort      = $q->{sort};

        if ( my $fn = $dispatch_table->{$sort} ) {
           my ( $tmp_search, $tmp_meta ) = $fn->( $direction );
           %search = ( %search, %{$tmp_search||{}} );
           %meta   = ( %meta,   %{$tmp_meta||{}} );
        } elsif ( $sort && $direction ) {
           my ( $tmp_search, $tmp_meta ) = $default->( $sort, $direction );
           %search = ( %search, %{$tmp_search||{}} );
           %meta   = ( %meta,   %{$tmp_meta||{}} );
        }

        return $self->search(\%search, \%meta);
     }

     # sort method in specific resultset
     sub controller_sort {
        my $self = shift;
        my $params = shift;
        return $self->_build_sort({
           first_name => sub {
              my $direction = shift;
              return {}, {
                 order_by => { "-$direction" => [qw{last_name first_name}] },
              };
           },
        }, sub {
           my $param = shift;
           my $direction = shift;
           return {}, {
              order_by => { "-$direction" => $param },
           };
        },$params);
     }

  simple_deletion
     $self->simple_deletion($c, $c->model('DB::Foo'));

    Deletes from the passed in resultset based on the following CGI
    parameter:

     to_delete - values of the ids of items to delete

    This is the only method that does not return a ResultSet. Instead it
    returns an arrayref of the id's that it deleted. If the ResultSet has
    has a multipk this will expect each tuple of PK's to be separated by
    commas.

    Note that this method uses the "$rs->delete" method, as opposed to
    "$rs->delete_all"

  simple_search
     my $searched_rs = $self->simple_search($c, $c->model('DB::Foo'));

    Searches the resultset based on all fields in the request, except for
    fields listed in "ignored_params". Searches with "$fieldname => { -like
    => "%$value%" }". If there are multiple values for a CGI parameter it
    will use all values via an "or".

  simple_sort
     my $sorted_rs = $self->simple_sort($c, $c->model('DB::Foo'));

    Sorts the passed in resultset based on the following CGI parameters:

     sort - field to sort by, defaults to primarky key
     dir  - direction to sort

CONFIG VARIABLES
    page_size
        Default size of a page. Defaults to 25.

    ignored_params
        ArrayRef of params that will be ignored in simple_search, defaults
        to:

         [qw{limit start sort dir _dc rm xaction}]

SEE ALSO
    Web::Util::DBIC::Paging, which this module is a thin wrapper around

CREDITS
    Thanks to Micro Technology Services, Inc. for funding the initial
    development of this module.

AUTHOR
    Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Arthur Axel "fREW" Schmidt.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Catalyst-TraitFor-Controller-DBIC-DoesPaging

1.001002  2013-05-31 23:45:39 CST6CDT
       - Stop using RT

1.001001  2012-11-17 15:47:43 CST6CDT
       - Put MetaYAML back in dist

1.001000  2010-03-10 16:32:59 CST6CDT
       - search and sort now call simple_search and simple_sort
         if resultset controller methods are undefined

1.00000   2010-03-08 20:51:00 CST6CDT
       - Change Version schema away from RJBS versioning
       - Fix tests to not use keywords for labels

0.093510  2009-12-17 00:12:03 CST6CDT
       - Add multipk deletion
       - Add aliases to a lot of things that could have been ambiguous before

0.093200 Mon Nov 16 21:08:00 CDT 2009
       - Fix required version for Catalyst::Model::DBIC::Schema
       - Fix required Catalyst version

0.093050 Sun Nov 01 01:09:00 CDT 2009
       - Add build_dep

0.093020 Thu Oct 29 19:16:00 CDT 2009
       - Rename dist because I'm THAT dumb

0.093010 Wed Oct 28 00:20:00 CDT 2009
       - Add basic code for multple values for a single CGI Parameter
         in simple_search

0.092980 Sun Oct 25 00:57:00 CDT 2009
       - Fix dependencies

0.092970 Sat Oct 24 02:47:00 CDT 2009
       - Initial Release



t/lib/TestApp.pm
package TestApp;
our $VERSION = '1.001000';

use strict;
use warnings;

use Catalyst::Runtime 5.80;

use parent qw/Catalyst/;
use Catalyst;

__PACKAGE__->config(
   name => 'TestApp',
   default_view => 'JSON',
   'View::JSON' => {
       expose_stash    => 'js',
    },
);

# Start the application
__PACKAGE__->setup();

1;


t/lib/TestApp/Schema.pm
package TestApp::Schema;
our $VERSION = '1.001000';
use parent 'DBIx::Class::Schema';

__PACKAGE__->load_namespaces();

"I am not a number, I'm a free man!";


t/lib/TestApp/Model/DB.pm
package TestApp::Model::DB;
our $VERSION = '1.001000';
use parent 'Catalyst::Model::DBIC::Schema';

__PACKAGE__->config(
   schema_class => 'TestApp::Schema',
   connect_info => { dsn => 'dbi:SQLite:dbname=test.db' }
);

1;


Build.PL

use strict;
use warnings;

use Module::Build;

my $build = Module::Build->new(
  module_name   => 'Catalyst::TraitFor::Controller::DBIC::DoesPaging',
  license       => 'perl',
  dist_abstract => "Helps\ you\ paginate\,\ search\,\ sort\,\ and\ more\ easily\ using\ DBIx\:\:Class",
  dist_name     => "Catalyst\-TraitFor\-Controller\-DBIC\-DoesPaging",
  dist_author   => [
"Arthur\ Axel\ \"fREW\"\ Schmidt\ \<frioux\+cpan\@gmail\.com\>",
  ],
  requires      => {
    "parent" => '0',
    "SQL::Translator" => '0.11002',
    "FindBin" => '0',
    "Carp" => '0',
    "Catalyst::Model::DBIC::Schema" => '0.29',
    "Test::More" => '0.88',
    "Catalyst::View::JSON" => '0',
    "Catalyst::Runtime" => '5.80000',
    "JSON" => '0',
    "Catalyst::Test" => '0',
    "Test::Deep" => '0',
    "DBIx::Class" => '0.08108',
  },
  script_files => [ qw() ],
);

$build->create_build_script;


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Helps you paginate, search, sort, and more easily using DBIx::Class",
  "AUTHOR" => "Arthur Axel \"fREW\" Schmidt <frioux+cpan\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Catalyst-TraitFor-Controller-DBIC-DoesPaging",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Catalyst::TraitFor::Controller::DBIC::DoesPaging",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Catalyst::Model::DBIC::Schema" => "0.29",
    "Catalyst::Runtime" => "5.80000",
    "Catalyst::Test" => 0,
    "Catalyst::View::JSON" => 0,
    "DBIx::Class" => "0.08108",
    "FindBin" => 0,
    "JSON" => 0,
    "SQL::Translator" => "0.11002",
    "Test::Deep" => 0,
    "Test::More" => "0.88",
    "parent" => 0
  },
  "TEST_REQUIRES" => {},
  "VERSION" => "1.001002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Catalyst/TraitFor/Controller/DBIC/DoesPaging.pm
package Catalyst::TraitFor::Controller::DBIC::DoesPaging;
{
  $Catalyst::TraitFor::Controller::DBIC::DoesPaging::VERSION = '1.001002';
}

# ABSTRACT: Helps you paginate, search, sort, and more easily using DBIx::Class

use Moose::Role;
use Web::Util::DBIC::Paging ':all' => { -prefix => '_' };

has ignored_params => (
   is => 'ro',
   isa => 'ArrayRef',
   default => sub { [qw{limit start sort dir _dc rm xaction}] },
);

has page_size => (
   is => 'ro',
   isa => 'Int',
   default => 25,
);

sub page_and_sort {
   _page_and_sort(c => $_[1], $_[2], { page_size => $_[0]->page_size })
}

sub paginate {
   _paginate(c => $_[1], $_[2], { page_size => $_[0]->page_size })
}

sub search { _search(c => $_[1], $_[2], { skip => $_[0]->ignored_params }) }

sub sort { _sort_rs(c => $_[1], $_[2]) }

sub simple_deletion { _simple_deletion(c => $_[1], $_[2]) }

sub simple_search {
   _simple_search(c => $_[1], $_[2], { skip => $_[0]->ignored_params });
}

sub simple_sort { _simple_sort(c => $_[1], $_[2]) }

1;

__END__

=pod

=head1 NAME

Catalyst::TraitFor::Controller::DBIC::DoesPaging - Helps you paginate, search, sort, and more easily using DBIx::Class

=head1 VERSION

version 1.001002

=head1 SYNOPSIS

 package MyApp::Controller::Foo;
 use Moose;
 BEGIN { extends 'Catalyst::Controller' }
 with 'Catalyst::TraitFor::Controller::DBIC::DoesPaging';

 sub people {
    my ($self, $c) = @_;
    my $people = $self->page_and_sort(
       $self->search( $self->model('DB::People') )
    );
    # ...
 }

=head1 DESCRIPTION

This module helps you to map various L<DBIx::Class> features to CGI parameters.
For the most part that means it will help you search, sort, and paginate with a
minimum of effort and thought.

=head1 METHODS

All methods take the context and a ResultSet as their arguments.  All methods
return a ResultSet.

=head2 page_and_sort

 my $result = $self->page_and_sort($c, $c->model('DB::Foo'));

This is a helper method that will first L</sort> your data and then L</paginate>
it.

=head2 paginate

 my $result = $self->paginate($c, $c->model('DB::Foo'));

Paginates the passed in resultset based on the following CGI parameters:

 start - first row to display
 limit - amount of rows per page

=head2 search

 my $searched_rs = $self->search($c, $c->model('DB::Foo'));

If the C<$resultset> has a C<controller_search> method it will call that method
on the passed in resultset with all of the CGI parameters.  I like to have this
method look something like the following:

 # Base search dispatcher, defined in MyApp::Schema::ResultSet
 sub _build_search {
    my $self           = shift;
    my $dispatch_table = shift;
    my $q              = shift;

    my %search = ();
    my %meta   = ();

    foreach ( keys %{$q} ) {
       if ( my $fn = $dispatch_table->{$_} and $q->{$_} ) {
          my ( $tmp_search, $tmp_meta ) = $fn->( $q->{$_} );
          %search = ( %search, %{$tmp_search||{}} );
          %meta   = ( %meta,   %{$tmp_meta||{}} );
       }
    }

    return $self->search(\%search, \%meta);
 }

 # search method in specific resultset
 sub controller_search {
    my $self   = shift;
    my $params = shift;
    return $self->_build_search({
       status => sub {
          return { 'repair_order_status' => shift }, {};
       },
       part_id => sub {
          return {
             'lineitems.part_id' => { -like => q{%}.shift( @_ ).q{%} }
          }, { join => 'lineitems' };
       },
    },$params);
 }

If the C<controller_search> method does not exist, this method will call
L</simple_search> instead.

=head2 sort

 my $result = $self->sort($c, $c->model('DB::Foo'));

Exactly the same as search, except calls C<controller_sort> or L</simple_sort>.
Here is how I use it:

 # Base sort dispatcher, defined in MyApp::Schema::ResultSet
 sub _build_sort {
    my $self = shift;
    my $dispatch_table = shift;
    my $default = shift;
    my $q = shift;

    my %search = ();
    my %meta   = ();

    my $direction = $q->{dir};
    my $sort      = $q->{sort};

    if ( my $fn = $dispatch_table->{$sort} ) {
       my ( $tmp_search, $tmp_meta ) = $fn->( $direction );
       %search = ( %search, %{$tmp_search||{}} );
       %meta   = ( %meta,   %{$tmp_meta||{}} );
    } elsif ( $sort && $direction ) {
       my ( $tmp_search, $tmp_meta ) = $default->( $sort, $direction );
       %search = ( %search, %{$tmp_search||{}} );
       %meta   = ( %meta,   %{$tmp_meta||{}} );
    }

    return $self->search(\%search, \%meta);
 }

 # sort method in specific resultset
 sub controller_sort {
    my $self = shift;
    my $params = shift;
    return $self->_build_sort({
       first_name => sub {
          my $direction = shift;
          return {}, {
             order_by => { "-$direction" => [qw{last_name first_name}] },
          };
       },
    }, sub {
       my $param = shift;
       my $direction = shift;
       return {}, {
          order_by => { "-$direction" => $param },
       };
    },$params);
 }

=head2 simple_deletion

 $self->simple_deletion($c, $c->model('DB::Foo'));

Deletes from the passed in resultset based on the following CGI parameter:

 to_delete - values of the ids of items to delete

This is the only method that does not return a ResultSet.  Instead it returns an
arrayref of the id's that it deleted.  If the ResultSet has has a multipk this will
expect each tuple of PK's to be separated by commas.

Note that this method uses the C<< $rs->delete >> method, as opposed to
C<< $rs->delete_all >>

=head2 simple_search

 my $searched_rs = $self->simple_search($c, $c->model('DB::Foo'));

Searches the resultset based on all fields in the request, except for fields
listed in C<ignored_params>.  Searches with
C<< $fieldname => { -like => "%$value%" } >>.  If there are multiple values for
a CGI parameter it will use all values via an C<or>.

=head2 simple_sort

 my $sorted_rs = $self->simple_sort($c, $c->model('DB::Foo'));

Sorts the passed in resultset based on the following CGI parameters:

 sort - field to sort by, defaults to primarky key
 dir  - direction to sort

=head1 CONFIG VARIABLES

=over 4

=item page_size

Default size of a page.  Defaults to 25.

=item ignored_params

ArrayRef of params that will be ignored in simple_search, defaults to:

 [qw{limit start sort dir _dc rm xaction}]

=back

=head1 SEE ALSO

L<Web::Util::DBIC::Paging>, which this module is a thin wrapper around

=head1 CREDITS

Thanks to Micro Technology Services, Inc. for funding the initial development
of this module.

=head1 AUTHOR

Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Arthur Axel "fREW" Schmidt.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
